### PR TITLE
DTSERWTHRE221 - Code updated for Transfer zero fund issue fix

### PR DESCRIPTION
**Pull request by nanbehera**
TransferFund Zero fund issue

### CODE REVIEW

File: Transfer/Sources/TransferAllFundsCell.swift [hunk 1]
```diff
@@ -85,7 +85,8 @@ final class TransferAllFundsCell: UITableViewCell {
                    currencyCode: String?) {
         guard let availableBalance = availableBalance,
             let currencyCode = currencyCode else {
-                availableFundsLabel.text = ""
+                availableFundsLabel.text = "Amount available: N/A"
```
Comment (jsaini-hw) [inline Transfer/Sources/TransferAllFundsCell.swift:88]: This should come from localization

File: Transfer/Sources/CreateTransferController.swift [hunk 2]
```diff
@@ -43,6 +43,7 @@ final class CreateTransferController: UITableViewController {
         (TransferButtonCell.self, TransferButtonCell.reuseIdentifier),
         (TransferNotesCell.self, TransferNotesCell.reuseIdentifier)
     ]
+    private var buttonSelection: Bool = false
```
Comment (jsaini-hw) [inline Transfer/Sources/CreateTransferController.swift:46]: No need of adding this. Move the handling of error.group != .business in presenter#createInitialTransfer()

File: Transfer/Sources/CreateTransferController.swift [hunk 3]
```diff
@@ -287,6 +288,7 @@ extension CreateTransferController {
 
     @objc
     private func tapTransfer(sender: UITapGestureRecognizer) {
+        buttonSelection = true
```
Comment (jsaini-hw) [inline Transfer/Sources/CreateTransferController.swift:291]: no need of tis

File: Transfer/Sources/CreateTransferController.swift [hunk 4]
```diff
@@ -376,7 +378,13 @@ extension CreateTransferController: CreateTransferView {
 
     func showError(_ error: HyperwalletErrorType, pageName: String, pageGroup: String, _ retry: (() -> Void)?) {
         let errorView = ErrorView(viewController: self, error: error, pageName: pageName, pageGroup: pageGroup)
-        errorView.show(retry)
+        print("show the error group *************", error.group)
+        if error.group == .business && buttonSelection == true {
+            errorView.show(retry)
+        } else if error.group != .business {
+            errorView.show(retry)
```
Comment (jsaini-hw) [inline Transfer/Sources/CreateTransferController.swift:385]: what's the difference in if and else logic? Also handle this in Presenter

**Review by jsaini-hw — CHANGES_REQUESTED** [review]

File: Transfer/Sources/TransferAllFundsCell.swift (hunk 1 above)
Comment (nanbehera) [reply Transfer/Sources/TransferAllFundsCell.swift:88]: Yes added translation for English. Code updated.Waiting for other language translations.

File: Transfer/Sources/CreateTransferController.swift (hunk 3 above)
Comment (nanbehera) [reply Transfer/Sources/CreateTransferController.swift]: Ok

File: Transfer/Sources/CreateTransferController.swift (hunk 2 above)
Comment (nanbehera) [reply Transfer/Sources/CreateTransferController.swift]: ok

File: Transfer/Sources/CreateTransferController.swift (hunk 4 above)
Comment (nanbehera) [reply Transfer/Sources/CreateTransferController.swift]: ok

File: Common/Sources/Localization/en.lproj/Localizable.strings [hunk 5]
```diff
@@ -249,6 +249,7 @@
 "mobileFXbuy" = "To";
 "mobileFXRateLabel" = "Exchange rate";
 "mobileAvailableBalance" = "Amount available %1$@%2$@ %3$@";
+"zeroAvailableBalance" = "Amount available: N/A";
```
Comment (jsaini-hw) [inline Common/Sources/Localization/en.lproj/Localizable.strings:252]: change it naAvailableBalance

File: Transfer/Sources/CreateTransferPresenter.swift [hunk 6]
```diff
@@ -325,8 +325,10 @@ final class CreateTransferPresenter {
             view.hideLoading()
             switch result {
             case .failure(let error):
+                if error.group != .business {
                 view.showError(error, pageName: strongSelf.pageName, pageGroup: strongSelf.pageGroup) {
```
Comment (jsaini-hw) [inline Transfer/Sources/CreateTransferPresenter.swift:329]: add indentation

Comment (jsaini-hw) [reply Transfer/Sources/CreateTransferPresenter.swift:329]: Update failing unit test for this case

**Review by jsaini-hw — CHANGES_REQUESTED** [review]
Travis CI is failing. Kindly look into it.

File: Common/Sources/Localization/en.lproj/Localizable.strings (hunk 5 above)
Comment (nanbehera) [reply Common/Sources/Localization/en.lproj/Localizable.strings:252]: Done

File: Transfer/Sources/CreateTransferPresenter.swift (hunk 6 above)
Comment (nanbehera) [reply Transfer/Sources/CreateTransferPresenter.swift:329]: ok

**Review by aparthibanpaypal — APPROVED** [review]

**Review by pdamodarannair — DISMISSED** [review]

**Comment by nanbehera** [comment]
Screenshot 2020-12-18 at 12 14 15 AM
Screenshot 2020-12-18 at 12 15 12 AM
UI and Unit tests

**Review by jsaini-hw — APPROVED** [review]

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 7]
```diff
@@ -207,13 +211,23 @@ class CreateTransferTests: XCTestCase {
...
+                           isAvailableBalancePresent: true)
+        } else {
+            mockView.stopOnError = false
+            initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
+            assertResponse(isShowErrorPerformed: false,
+                           transferSourceCellConfigurationsCount: 3,
+                           transferSourceType: .user,
+                           selectedTransferDestination: true,
+                           isAvailableBalancePresent: true)
+        }
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:230]: why enhance  this? Create a new test case for unexpected error

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 8]
```diff
@@ -248,13 +262,23 @@ class CreateTransferTests: XCTestCase {
...
+                           isAvailableBalancePresent: false)
+        } else {
+            mockView.stopOnError = false
+            initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
+            assertResponse(isShowErrorPerformed: false,
+                           transferSourceCellConfigurationsCount: 2,
+                           transferSourceType: .prepaidCard,
+                           selectedTransferDestination: true,
+                           isAvailableBalancePresent: false)
+        }
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:281]: same as above?

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 7 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:230]: ok

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 9]
```diff
@@ -211,18 +211,26 @@ class CreateTransferTests: XCTestCase {
...
                            isAvailableBalancePresent: true)
-        } else {
-            mockView.stopOnError = false
-            initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
-            assertResponse(isShowErrorPerformed: false,
+        }
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_walletModel_ErrorResponseforunexpectedError() {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:226]: Update testLoadCreateTransfer_showAllAvailableSources_walletModel_ErrorResponseforunexpectedError to testLoadCreateTransfer_showAllAvailableSources_walletModel_errorResponseforunexpectedError

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 10]
```diff
@@ -262,18 +270,26 @@ class CreateTransferTests: XCTestCase {
...
                            isAvailableBalancePresent: false)
-        } else {
-            mockView.stopOnError = false
-            initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
-            assertResponse(isShowErrorPerformed: false,
+        }
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_ErrorResponseforunexpectedError() {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:285]: testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_ErrorResponseforunexpectedError() -> testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_errorResponseforunexpectedError()

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 9 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: Updated

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 10 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: Updated

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 11]
```diff
@@ -12,6 +12,7 @@ class CreateTransferTests: XCTestCase {
     private var presenter: CreateTransferPresenter!
     private var mockView = MockCreateTransferView()
     private let clientTransferId = UUID().uuidString
+    private var error: HyperwalletErrorType!
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:15]: you don't need this

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 12]
```diff
@@ -207,13 +211,31 @@ class CreateTransferTests: XCTestCase {
...
         HyperwalletTestHelper.programModel = .walletModel
         Hyperwallet.setup(HyperwalletTestHelper.authenticationProvider)
-        mockView.stopOnError = true
-        initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
-        assertResponse(isShowErrorPerformed: true,
-                       transferSourceCellConfigurationsCount: 3,
-                       transferSourceType: .user,
-                       selectedTransferDestination: true,
-                       isAvailableBalancePresent: false)
+        if error?.group == .business {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:214]: no need of checking error.group, code takes care of tat. Unit test should not involve any business logic

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 13]
```diff
@@ -207,13 +211,31 @@ class CreateTransferTests: XCTestCase {
...
+            assertResponse(isShowErrorPerformed: false,
+                           transferSourceCellConfigurationsCount: 3,
+                           transferSourceType: .user,
+                           selectedTransferDestination: true,
+                           isAvailableBalancePresent: true)
+        }
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_walletModel_errorResponseforunexpectedError() {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:226]: testLoadCreateTransfer_showAllAvailableSources_walletModel_errorResponseforunexpectedError-> 
testLoadCreateTransfer_showAllAvailableSources_walletModel_errorResponseForUnexpectedError

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 14]
```diff
@@ -248,13 +270,31 @@ class CreateTransferTests: XCTestCase {
...
         HyperwalletTestHelper.programModel = .pay2CardModel
         Hyperwallet.setup(HyperwalletTestHelper.authenticationProvider)
-        mockView.stopOnError = true
-        initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
-        assertResponse(isShowErrorPerformed: true,
-                       transferSourceCellConfigurationsCount: 2,
-                       transferSourceType: .prepaidCard,
-                       selectedTransferDestination: true,
-                       isAvailableBalancePresent: false)
+        if error?.group == .business {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:273]: no need to check this

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 15]
```diff
@@ -248,13 +270,31 @@ class CreateTransferTests: XCTestCase {
...
+            assertResponse(isShowErrorPerformed: false,
+                           transferSourceCellConfigurationsCount: 2,
+                           transferSourceType: .prepaidCard,
+                           selectedTransferDestination: true,
+                           isAvailableBalancePresent: false)
+        }
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_errorResponseforunexpectedError() {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:285]: testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_errorResponseforunexpectedError -> 
testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_errorResponseForUnexpectedError

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 16]
```diff
@@ -248,13 +270,31 @@ class CreateTransferTests: XCTestCase {
...
+        Hyperwallet.clearInstance()
+        HyperwalletTestHelper.programModel = .pay2CardModel
+        Hyperwallet.setup(HyperwalletTestHelper.authenticationProvider)
+        if error?.group != .business {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:289]: no need to check this

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 17]
```diff
@@ -207,13 +211,31 @@ class CreateTransferTests: XCTestCase {
...
+                           isAvailableBalancePresent: true)
+        }
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_walletModel_errorResponseforunexpectedError() {
+        Hyperwallet.clearInstance()
+        HyperwalletTestHelper.programModel = .walletModel
+        Hyperwallet.setup(HyperwalletTestHelper.authenticationProvider)
+        if error?.group != .business {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:230]: no need to check this

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 11 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:15]: ok

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 12 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:214]: ok

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 13 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:226]: ok

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 16 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: ok

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 17 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: ok

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 15 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:278]: updated

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 14 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: removed

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 18]
```diff
@@ -737,8 +763,6 @@ class MockCreateTransferView: CreateTransferView {
         isHideLoadingPerformed = false
         isNotifyTransferCreatedPerformed = false
         isShowCreateTransferPerformed = false
-        isShowErrorPerformed = false
-        isShowAlertPerformed = false
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:741]: Why did you remove this?

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 19]
```diff
@@ -248,8 +265,22 @@ class CreateTransferTests: XCTestCase {
...
         initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
+        assertResponse(isShowErrorPerformed: false,
+                       transferSourceCellConfigurationsCount: 2,
+                       transferSourceType: .prepaidCard,
+                       selectedTransferDestination: true,
+                       isAvailableBalancePresent: false)
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_pay2CardModel_errorResponseforunexpectedError() {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:278]: it's still not updated

File: Transfer/Tests/CreateTransferPresenterTests.swift [hunk 20]
```diff
@@ -207,13 +210,27 @@ class CreateTransferTests: XCTestCase {
...
         initializePresenter(createTransferResult: .failure, showAllAvailableSources: true)
+        assertResponse(isShowErrorPerformed: false,
+                       transferSourceCellConfigurationsCount: 3,
+                       transferSourceType: .user,
+                       selectedTransferDestination: true,
+                       isAvailableBalancePresent: true)
+        Hyperwallet.clearInstance()
+    }
+
+    func testLoadCreateTransfer_showAllAvailableSources_walletModel_errorResponseforunexpectedError() {
```
Comment (jsaini-hw) [inline Transfer/Tests/CreateTransferPresenterTests.swift:223]: it's still not updated

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 18 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:741]: fixed

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 19 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:278]: yeah removed error group check logic

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 20 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift:223]: yeah removed error group check logic

Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: naming issue fixed

File: Transfer/Tests/CreateTransferPresenterTests.swift (hunk 19 above)
Comment (nanbehera) [reply Transfer/Tests/CreateTransferPresenterTests.swift]: naming issue fixed

**Review by pdamodarannair — APPROVED** [review]

**Review by jsaini-hw — APPROVED** [review]
